src/mlu/user_data_proxy.rs

```diff
@@ -0,0 +1,45 @@
+use std::marker::PhantomData;
+
+use mlua::{AnyUserData, Error, Lua, ToLua, UserData};
+
+use crate::{TypeBody, TypeName};
+
+/// A userdata which can be used as a static proxy
+pub trait StaticUserdata: UserData + 'static {}
+impl<T: UserData + 'static> StaticUserdata for T {}
+
+/// A newtype storing userdata created via [`mlua::Lua::create_proxy`].
+///
+/// if `T` implements TypeName or TypeBody the implementations are forwarded to this type.
+pub struct UserDataProxy<'lua, T: StaticUserdata> {
+    user_data: AnyUserData<'lua>,
+    ph_: PhantomData<T>,
+}
+
+impl<'lua, T: StaticUserdata> UserDataProxy<'lua, T> {
+    /// Creates a new UserDataProxy
+    pub fn new(lua: &'lua Lua) -> Result<Self, Error> {
+        Ok(Self {
+            user_data: lua.create_proxy::<T>()?,
+            ph_: Default::default(),
+        })
+    }
+}
+
+impl<T: StaticUserdata + TypeName> TypeName for UserDataProxy<'_, T> {
+    fn get_type_parts() -> std::borrow::Cow<'static, [crate::NamePart]> {
+        T::get_type_parts()
+    }
+}
+
+impl<T: StaticUserdata + TypeBody> TypeBody for UserDataProxy<'_, T> {
+    fn get_type_body() -> crate::TypeGenerator {
+        T::get_type_body()
```

Comment:
This means that the proxy has _all_ members of the type that it is proxying, but it should only have those that are static

Reply:
See my previous comment, and in light of it, maybe it doesn't make sense for this to have TypeBody at all?